recipes/full_finetune_distributed.py

```diff
@@ -199,21 +204,101 @@ def __init__(self, cfg: DictConfig) -> None:
         self.total_epochs = cfg.epochs
         self.max_steps_per_epoch = cfg.max_steps_per_epoch
         self.global_step = 0
+        self._dcp_checkpointer = None
+
+    def get_dcp_checkpointer(self):
+        if not self._dcp_checkpointer:
+            self._dcp_checkpointer = DistributedCheckpointer(
+                checkpoint_dir=self._checkpointer._checkpoint_dir,
+                output_dir=self._checkpointer._output_dir,
+            )
+
+        return self._dcp_checkpointer
 
     def load_checkpoint(self, cfg_checkpointer: DictConfig) -> Dict[str, Any]:
         """
-        Extract the checkpoint state from file and validate. If resume_from_checkpoint
-        is True, this also includes the recipe state.
+        Extract the checkpoint state from file and validate.
+        If resume_from_checkpoint is True, this also includes the recipe state.
         """
+        # If async checkpointing is enabled, set the should_load_recipe_state as false for the checkpointer.
+        # We will use DistributedCheckpointer instead to load the intermediate checkpoints which includes the
+        # recipe state as well.
+        should_load_recipe_state: bool = (
+            False if self._enable_async_checkpointing else self._resume_from_checkpoint
+        )
+
         self._checkpointer = config.instantiate(
             cfg_checkpointer,
-            resume_from_checkpoint=self._resume_from_checkpoint,
+            should_load_recipe_state=should_load_recipe_state,
         )
-        checkpoint_dict = self._checkpointer.load_checkpoint()
 
-        if self._resume_from_checkpoint:
+        return self._checkpointer.load_checkpoint()
+
+    def resume_from_checkpoint(self, checkpoint_dict: Dict[str, any]) -> Dict[str, Any]:
```

Comment:
Personally, i would prefer if we broke it down into:

```
def resume_from_checkpoint_async
	...

def resume_from_checkpoint_sync
	...
```

Not a big fan of:
```
if:
	100 lines
else:
	...
```

Do you think that it would make the logic more discrete and easier to maintain?